### PR TITLE
18026 Add allowable actions for AGM location change

### PR DESCRIPTION
**Pull request by leodube-aot**
*Issue #:* /bcgov/entity#18026

*Description of changes:*
-  Added Allowable Change item for filing type = `agmLocationChange`
- Applicable only to company types - BC Ltd., BEN, ULC, CCC
- Filing is not allowed when one or more of following is true for business
  - Business is in "Not in Good Standing" state
  - Business is frozen
  - Business has a pending or draft filing
- Created Unit Tests

Note: The unit tests are currently blocked by tickets in progress.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/tests/unit/services/test_authorization.py [hunk 1]
```diff
@@ -25,9 +25,9 @@
 
 import pytest
 from flask import jsonify
-from registry_schemas.example_data import ALTERATION_FILING_TEMPLATE, ANNUAL_REPORT, CORRECTION_AR, \
-    CHANGE_OF_REGISTRATION_TEMPLATE, RESTORATION, FILING_TEMPLATE, DISSOLUTION, PUT_BACK_ON, CONTINUATION_IN, \
-    CONSENT_CONTINUATION_OUT, CONTINUATION_OUT
+from registry_schemas.example_data import AGM_LOCATION_CHANGE, ALTERATION_FILING_TEMPLATE, ANNUAL_REPORT, \
+    CORRECTION_AR, CHANGE_OF_REGISTRATION_TEMPLATE, RESTORATION, FILING_TEMPLATE, DISSOLUTION, PUT_BACK_ON, \
+    CONTINUATION_IN, CONSENT_CONTINUATION_OUT, CONTINUATION_OUT
```
Comment (leodube-aot) [inline legal-api/tests/unit/services/test_authorization.py:31]: Currently waiting on bcgov/entity#17941 to get merged to use the `AGM_LOCATION_CHANGE` example data.

File: legal-api/tests/unit/services/test_authorization.py [hunk 2]
```diff
@@ -159,6 +160,7 @@ class FilingKey(str, Enum):
                                     'name': 'registrarsNotation'},
     FilingKey.REGISTRARS_ORDER: {'displayName': "Registrar's Order", 'feeCode': 'NOFEE', 'name': 'registrarsOrder'},
     FilingKey.SPECIAL_RESOLUTION: {'displayName': 'Special Resolution', 'feeCode': 'SPRLN', 'name': 'specialResolution'},
+    FilingKey.AGM_LOCATION_CHANGE: {'displayName': '', 'feeCode': '', 'name': ''},
     FilingKey.ALTERATION: {'displayName': 'Alteration', 'feeCode': 'ALTER', 'name': 'alteration'},
```
Comment (leodube-aot) [inline legal-api/tests/unit/services/test_authorization.py:164]: Also waiting on https://github.com/bcgov/entity/issues/17940 to set and test the expected filing data.

File: legal-api/src/legal_api/services/authz.py [hunk 3]
```diff
@@ -267,6 +278,17 @@ def has_roles(jwt: JwtManager, roles: List[str]) -> bool:
...
+                'legalTypes': ['BC', 'BEN', 'ULC', 'CC'],
+                'blockerChecks': {
+                    'business': [
+                        BusinessBlocker.DEFAULT,
+                        BusinessBlocker.BUSINESS_FROZEN,
+                        BusinessBlocker.DRAFT_PENDING,
+                        BusinessBlocker.NOT_IN_GOOD_STANDING
+                    ]
+                }
+            },
```
Comment (leodube-aot) [inline legal-api/src/legal_api/services/authz.py:291]: Just checking, but we wanted to allow both staff and general, correct?

Comment (argush3) [reply legal-api/src/legal_api/services/authz.py:291]: I think so.  But can you confirm with Mihai?

File: legal-api/src/legal_api/services/authz.py [hunk 4]
```diff
@@ -118,6 +118,17 @@ def has_roles(jwt: JwtManager, roles: List[str]) -> bool:
             'adminFreeze': {
                 'legalTypes': ['SP', 'GP', 'CP', 'BC', 'BEN', 'CC', 'ULC'],
             },
+            'agmLocationChange': {
+                'legalTypes': ['BC', 'BEN', 'ULC', 'CC'],
+                'blockerChecks': {
+                    'business': [
+                        BusinessBlocker.DEFAULT,
+                        BusinessBlocker.BUSINESS_FROZEN,
+                        BusinessBlocker.DRAFT_PENDING,
```
Comment (argush3) [inline legal-api/src/legal_api/services/authz.py:127]: business frozen and draft pending is already contained in default so we can remove these two.

File: legal-api/tests/unit/services/test_authorization.py
Comment (argush3) [inline legal-api/tests/unit/services/test_authorization.py:1]: I'm a bit surprised we never wrote any not in good standing unit tests for allowable actions.  

Can you add some?  Probably something similar to [test_get_allowed_filings_blocker_admin_freeze](https://github.com/bcgov/lear/blob/2925473ba764bc05863bab504b283f81beb9d7ab/legal-api/tests/unit/services/test_authorization.py#L1041C1-L1043C1)

Comment (leodube-aot) [reply legal-api/tests/unit/services/test_authorization.py:1]: Yup I can add some!

File: legal-api/requirements.txt [hunk 5]
```diff
@@ -59,4 +59,4 @@ minio==7.0.2
 PyPDF2==1.26.0
 reportlab==3.6.12
 html-sanitizer==1.9.3
-git+https://github.com/bcgov/business-schemas.git@2.18.11#egg=registry_schemas
+git+https://github.com/bcgov/business-schemas.git@2.18.12#egg=registry_schemas
```
Comment (argush3) [inline legal-api/requirements.txt:62]: we'll also need to update the schema [here](https://github.com/bcgov/lear/blob/main/legal-api/requirements/bcregistry-libraries.txt).

File: legal-api/src/legal_api/services/authz.py (hunk 3 above)
Comment (leodube-aot) [reply legal-api/src/legal_api/services/authz.py:281]: Confirmed with Mihai that both will be used.

Comment (argush3) [reply legal-api/src/legal_api/services/authz.py:281]: ok, thanks for checking!

**Review by vysakh-menon-aot — APPROVED** [review]

File: legal-api/tests/unit/services/test_authorization.py [hunk 6]
```diff
@@ -1062,6 +1083,60 @@ def mock_auth(one, two):  # pylint: disable=unused-argument; mocks of library me
...
+            business = None
+            if business_exists:
+                identifier = (f'BC{random.SystemRandom().getrandbits(0x58)}')[:9]
+                business = factory_business(identifier=identifier,
+                                            entity_type=legal_type,
+                                            state=state)
+                business.good_standing = False
+            filing_types = get_allowed_filings(business, state, legal_type, jwt)
+            print("test:")
+            print(filing_types)
```
Comment (argush3) [inline legal-api/tests/unit/services/test_authorization.py:1137]: can we remove the print statements when you are good?

File: legal-api/tests/unit/services/test_authorization.py [hunk 7]
```diff
@@ -1062,6 +1083,60 @@ def mock_auth(one, two):  # pylint: disable=unused-argument; mocks of library me
...
+        monkeypatch.setattr('flask.request.headers.get', mock_auth)
+
+        for legal_type in legal_types:
+            business = None
+            if business_exists:
+                identifier = (f'BC{random.SystemRandom().getrandbits(0x58)}')[:9]
+                business = factory_business(identifier=identifier,
+                                            entity_type=legal_type,
+                                            state=state)
+                business.good_standing = False
```
Comment (argush3) [inline legal-api/tests/unit/services/test_authorization.py:1134]: i think you'll need to patch the return value for this as it's a getter.

File: legal-api/tests/unit/services/test_authorization.py [hunk 8]
```diff
@@ -1062,6 +1083,60 @@ def mock_auth(one, two):  # pylint: disable=unused-argument; mocks of library me
...
+        ('staff_historical_llc', True, Business.State.HISTORICAL, ['LLC'], 'staff', [STAFF_ROLE], []),
+        ('staff_historical_firms', True, Business.State.HISTORICAL, ['SP', 'GP'], 'staff', [STAFF_ROLE],[]),
+
+        # historical business - general user
+        ('general_user_historical_cp', True, Business.State.HISTORICAL, ['CP'], 'general', [BASIC_USER], []),
+        ('general_user_historical_corps', True, Business.State.HISTORICAL, ['BC', 'BEN', 'CC', 'ULC'], 'general',
+         [BASIC_USER], []),
+        ('general_user_historical_llc', True, Business.State.HISTORICAL, ['LLC'], 'general', [BASIC_USER], []),
+        ('general_user_historical_firms', True, Business.State.HISTORICAL, ['SP', 'GP'], 'general', [BASIC_USER], []),
+    ]
```
Comment (argush3) [inline legal-api/tests/unit/services/test_authorization.py:1113]: for a lot of these tests i still expect some allowable actions returned even if the business is not in good standing

File: legal-api/tests/unit/services/test_authorization.py (hunk 7 above)
Comment (leodube-aot) [reply legal-api/tests/unit/services/test_authorization.py:1134]: Good call, I was looking into setting the `last_ar_date` to a value to make it not in good standing but patching it should be easier

File: legal-api/tests/unit/services/test_authorization.py (hunk 8 above)
Comment (leodube-aot) [reply legal-api/tests/unit/services/test_authorization.py:1113]: Yeah I just haven't added the expected actions yet. Having trouble running the tests on my local so I'm just trying to get it working in the PR, then adding in the expected actions.

**Review by argush3 — APPROVED** [review]

**Pull request merged**